Que tal praticar?
Este é o projeto que fecha sua jornada no curso e simula o dia a dia real de um Cientista de Dados: você vai receber três bases de dados diferentes, cada uma escondendo uma oportunidade de projeto, e será sua responsabilidade identificar o problema, definir a abordagem e construir a solução do zero — exatamente como acontece quando um time de dados recebe uma base "crua" de um cliente ou de outra área da empresa. Aqui entram todas as competências desenvolvidas ao longo do curso: da preparação e exploração dos dados até a escolha, treinamento e avaliação criteriosa do modelo. Mais do que rodar um algoritmo, o objetivo é demonstrar que você consegue conduzir um projeto de dados completo, com julgamento técnico e comunicação clara dos resultados.

O que você vai desenvolver com esse exercício
Capacidade de identificar e definir um problema de negócio a partir de uma base de dados bruta
Condução completa de uma Análise Exploratória de Dados (EDA), com múltiplas técnicas
Preparação de dados para modelagem, sem vazamento de dados (data leakage)
Treinamento e avaliação criteriosa de um modelo de classificação binária
Interpretação de resultados conectada ao problema de negócio, não apenas à métrica
O contexto do desafio
Serão disponibilizadas três bases de dados diferentes, cada uma contendo uma oportunidade de projeto que você mesmo deve identificar. Você escolherá uma dessas bases para trabalhar e aplicará todos os conceitos vistos ao longo do curso — desde a preparação do dado e o desenvolvimento do problema a ser solucionado, até a escolha do modelo e sua avaliação final.

O que deve ser entregue
Definição explícita do problema de negócio, do tipo de tarefa de Machine Learning e da variável-alvo escolhida
Análise Exploratória de Dados completa, com seção textual consolidando os principais insights
Base de dados higienizada e preparada para modelagem (tratamento de ausentes, tipagem, encoding)
Separação treino/teste feita antes das transformações sensíveis, sem vazamento de dados
Modelo de classificação binária treinado e avaliado com métricas apropriadas, com interpretação do resultado

Passo a passo
Escolha uma das três bases de dados disponibilizadas e identifique a oportunidade de projeto nela contida.
Defina explicitamente o problema de negócio, o tipo de tarefa de Machine Learning (classificação binária) e a variável-alvo.
Aplique pelo menos 5 técnicas distintas de Análise Exploratória de Dados sobre os dados reais da base escolhida.
Escreva uma seção textual consolidando os principais insights obtidos na fase exploratória.
Higienize e prepare os dados para a modelagem: trate valores ausentes, ajuste a tipagem e codifique as variáveis categóricas (encoding).
Separe os dados em treino e teste antes de aplicar qualquer transformação sensível, evitando vazamento de dados (data leakage).
Treine pelo menos um modelo apropriado para a tarefa de classificação binária definida.
Avalie o modelo com as métricas corretas para o problema e interprete o resultado — não basta apenas imprimir os números, é preciso explicar o que eles significam para o negócio.

Requisitos obrigatórios
A avaliação confere se o problema de negócio e a variável-alvo foram claramente definidos, se a EDA foi conduzida com profundidade (mínimo de 5 técnicas e insights consolidados), se os dados foram preparados sem vazamento de dados e se o modelo foi treinado e avaliado com interpretação real dos resultados. Critérios de excelência avaliam se você reconheceu e tratou alguma armadilha estrutural própria da base escolhida, se a EDA trouxe interpretação crítica e estatística além dos gráficos, se houve comparação entre múltiplos modelos e/ou validação cruzada com otimização de hiperparâmetros, se a importância das variáveis foi conectada ao domínio do problema, e se o resultado foi convertido em uma recomendação de negócio (Data Storytelling). Não é obrigatório usar as três bases disponibilizadas — apenas uma delas, escolhida e justificada por você.

Formato da entrega
Upload de imagens (prints do notebook com todas as etapas, da EDA à avaliação final do modelo)
Até 10 arquivos de imagem

Definição explícita do problema de negócio, do tipo de tarefa de Machine Learning e da variável-alvo escolhida.
Identificação precoce dos pacietes que apresentam a doença Hipotireodismo. 
As duas técnicas escolhidas foram Xgboost(cria várias árvores de decisão para encontrar o melhor caminho) e SVM(para buscar uma linha que melhor separe as classes) por se tratar de um problema problema de classificação.
A variável alvo é binaryClass por ser justamente a classificação final do caso, indicando a categoria diagnóstica. 
 

O diagnóstico precoce é importante porque manter o corpo em um estado de metabolismo acelerado por muito tempo sem tratamento traz reflexos negativos dada a sobrecarga a qual os órgãos estão sujeitos. Os sintomas iniciais podem ser leves e inespecíficos por isso é importante investigar criteriosamente.

!['Ciclo Funcionamento Tireóide'](https://lh4.googleusercontent.com/CR1UW9CVfU9TcPEP0xI5aXtzVL1OCQ3kOGJGnwNSWBuSgK86GV2FEmnYia_L9Q73BgeiOi26yrDMpmThr-e_ZVtilQYVdP48JE7WqsHN0KQjV0jA0eSub7a3WmtzIi5ti-EJ-Oc=s1600)

In [1]:
import pandas as pd

In [12]:
df = pd.read_csv("Base_M43_Pratique_Hypothyroid.csv", delimiter =',')

In [13]:
df[["TBG"]].head(10)
#confirmando que n tem dado

,TBG
0,?
1,?
2,?
3,?
4,?
5,?
6,?
7,?
8,?
9,?


In [14]:
print(df.head(15))

   age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0   41   F            f                  f                         f    f   
1   23   F            f                  f                         f    f   
2   46   M            f                  f                         f    f   
3   70   F            t                  f                         f    f   
4   70   F            f                  f                         f    f   
5   18   F            t                  f                         f    f   
6   59   F            f                  f                         f    f   
7   80   F            f                  f                         f    f   
8   66   F            f                  f                         f    f   
9   68   M            f                  f                         f    f   
10  84   F            f                  f                         f    f   
11  67   F            t                  f                         f    f   

In [15]:
print(df.columns.tolist())

['age', 'sex', 'on thyroxine', 'query on thyroxine', 'on antithyroid medication', 'sick', 'pregnant', 'thyroid surgery', 'I131 treatment', 'query hypothyroid', 'query hyperthyroid', 'lithium', 'goitre', 'tumor', 'hypopituitary', 'psych', 'TSH measured', 'TSH', 'T3 measured', 'T3', 'TT4 measured', 'TT4', 'T4U measured', 'T4U', 'FTI measured', 'FTI', 'TBG measured', 'TBG', 'referral source', 'binaryClass']


In [16]:
| Nome da coluna              | Explicação                                                            |
| --------------------------- | --------------------------------------------------------------------- |
| `on thyroxine`              | Indica se o paciente está tomando tiroxina (T4).                      |
| `query on thyroxine`        | Indica se foi investigado se o paciente usa tiroxina.                 |
| `on antithyroid medication` | Indica se o paciente usa medicamento antitireoidiano.                 |
| `thyroid surgery`           | Indica se o paciente já fez cirurgia da tireoide.                     |
| `I131 treatment`            | Indica se o paciente recebeu tratamento com iodo radioativo (I-131).  |
| `query hypothyroid`         | Indica se houve investigação de hipotireoidismo.                      |
| `query hyperthyroid`        | Indica se houve investigação de hipertireoidismo.                     |
| `lithium`                   | Indica se o paciente usa lítio, que pode afetar a tireoide.           |
| `goitre`                    | Indica presença de bócio, aumento da tireoide.                        |
| `hypopituitary`             | Indica presença/suspeita de hipopituitarismo.                         |
| `psych`                     | Indica histórico ou presença de condição psiquiátrica.                |
| `TSH measured`              | Indica se o exame de TSH foi realizado.                               |
| `TSH`                       | Valor do hormônio estimulante da tireoide.                            |
| `T3 measured`               | Indica se o exame de T3 foi realizado.                                |
| `T3`                        | Valor do hormônio triiodotironina.                                    |
| `TT4 measured`              | Indica se o exame de T4 total foi realizado.                          |
| `TT4`                       | Valor do T4 total no sangue.                                          |
| `T4U measured`              | Indica se o exame de T4U foi realizado.                               |
| `T4U`                       | Medida relacionada à ligação dos hormônios tireoidianos às proteínas. |
| `FTI measured`              | Indica se o índice de tiroxina livre foi calculado.                   |
| `FTI`                       | Índice que estima a quantidade de tiroxina livre disponível.          |
| `TBG measured`              | Indica se a TBG foi medida.                                           |
| `TBG`                       | Proteína que transporta principalmente T4 e T3 no sangue.             |
| `referral source`           | Indica a origem do encaminhamento do paciente.                        |
| `binaryClass`               | Classificação final do caso, indicando a categoria diagnóstica.       |


SyntaxError: invalid syntax (2221179022.py, line 1)

Análise Exploratória de Dados completa, com seção textual consolidando os principais insights Base de dados higienizada e preparada para modelagem (tratamento de ausentes, tipagem, encoding).

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3772 entries, 0 to 3771
Data columns (total 30 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   age                        3772 non-null   str  
 1   sex                        3772 non-null   str  
 2   on thyroxine               3772 non-null   str  
 3   query on thyroxine         3772 non-null   str  
 4   on antithyroid medication  3772 non-null   str  
 5   sick                       3772 non-null   str  
 6   pregnant                   3772 non-null   str  
 7   thyroid surgery            3772 non-null   str  
 8   I131 treatment             3772 non-null   str  
 9   query hypothyroid          3772 non-null   str  
 10  query hyperthyroid         3772 non-null   str  
 11  lithium                    3772 non-null   str  
 12  goitre                     3772 non-null   str  
 13  tumor                      3772 non-null   str  
 14  hypopituitary              3772 non

Os dados são todos do tipo string mesmo os que são numéricos, para algumas colunas ele usa p e n em outras t e f. Os dados ausentes não estão como 0.São 61 dados duplicados o que é 1,62% dos RangeIndex: 3772 entries, 0 to 3771 ded linhas então optei por remover.

In [18]:
df.isnull().sum()

age                          0
sex                          0
on thyroxine                 0
query on thyroxine           0
on antithyroid medication    0
sick                         0
pregnant                     0
thyroid surgery              0
I131 treatment               0
query hypothyroid            0
query hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH measured                 0
TSH                          0
T3 measured                  0
T3                           0
TT4 measured                 0
TT4                          0
T4U measured                 0
T4U                          0
FTI measured                 0
FTI                          0
TBG measured                 0
TBG                          0
referral source              0
binaryClass                  0
dtype: int64

In [19]:
df.eq('?').sum()

age                             1
sex                           150
on thyroxine                    0
query on thyroxine              0
on antithyroid medication       0
sick                            0
pregnant                        0
thyroid surgery                 0
I131 treatment                  0
query hypothyroid               0
query hyperthyroid              0
lithium                         0
goitre                          0
tumor                           0
hypopituitary                   0
psych                           0
TSH measured                    0
TSH                           369
T3 measured                     0
T3                            769
TT4 measured                    0
TT4                           231
T4U measured                    0
T4U                           387
FTI measured                    0
FTI                           385
TBG measured                    0
TBG                          3772
referral source                 0
binaryClass   

In [9]:
df.duplicated().sum()

np.int64(3770)

In [10]:
df.drop_duplicates(inplace=True)

In [11]:
df.duplicated().sum()

np.int64(0)

Separei em numérica e categórica para facilitar a manipulação

In [10]:
colunas_numericas = [
    'age',
    'TSH',
    'T3',
    'TT4',
    'T4U',
    'FTI',
    'TBG'
]

In [11]:
df[[ 'age',
    'TSH',
    'T3',
    'TT4',
    'T4U',
    'FTI',
    'TBG'
]].head(10) 
#conferindo p ver se é número

,age,TSH,T3,TT4,T4U,FTI,TBG
0,41,1.3,2.5,125,1.14,109,?
1,23,4.1,2,102,?,?,?
2,46,0.98,?,109,0.91,120,?
3,70,0.16,1.9,175,?,?,?
4,70,0.72,1.2,61,0.87,70,?
5,18,0.03,?,183,1.3,141,?
6,59,?,?,72,0.92,78,?
7,80,2.2,0.6,80,0.7,115,?
8,66,0.6,2.2,123,0.93,132,?
9,68,2.4,1.6,83,0.89,93,?


In [12]:
df[colunas_numericas] = df[colunas_numericas].apply(pd.to_numeric, errors='coerce')
#o coerce é p transformar ? em Nan sem dar erro

In [43]:
df[colunas_numericas].info()

KeyError: "None of [Index(['age', 'TSH', 'T3', 'TT4', 'T4U', 'FTI', 'TBG'], dtype='str')] are in the [columns]"

In [17]:
df[colunas_numericas].isnull().sum()

age       1
TSH     369
T3      769
TT4     231
T4U     387
FTI     385
TBG    3772
dtype: int64

In [18]:
#completando os nulos com a média para não perder os dados 
df[colunas_numericas] = df[colunas_numericas].fillna(
    df[colunas_numericas].mean()
)

In [19]:
df[colunas_numericas].isnull().sum()

age       0
TSH       0
T3        0
TT4       0
T4U       0
FTI       0
TBG    3772
dtype: int64

In [20]:
df['TBG'].unique()

array([nan])

In [ ]:
colunas_booleanas = [
    'on thyroxine',
    'query on thyroxine',
    'on antithyroid medication',
    'sick',
    'pregnant',
    'thyroid surgery',
    'I131 treatment',
    'query hypothyroid',
    'query hyperthyroid',
    'lithium',
    'goitre',
    'tumor',
    'hypopituitary',
    'psych',
    'TSH measured',
    'T3 measured',
    'TT4 measured',
    'T4U measured',
    'FTI measured',
    'TBG measured'
]

In [ ]:
colunas_categoricas = [
    'sex',
    'referral source']

A separação entre categórica e binária foi feita porque a primeira traz como informação uma categoria(responde o que é) e a segunda apenas um sim/não. A variável target(binaryClass) foi mantida isolada para tratamento cauteloso.

In [19]:
#transformar colunas para numérico
#fazer restrição para idade
#Transformar td p 0 e 1 TEM QUE VEM O QUE TEM EM CADA ANTES DE TROCAR!!
#tratar nulos
#Pegar médias e dps
#fazer padronização

Separação treino/teste feita antes das transformações sensíveis, sem vazamento de dados Modelo de classificação binária treinado e avaliado com métricas apropriadas, com interpretação do resultado.

Avalie o modelo com as métricas corretas para o problema e interprete o resultado.